src/main/scala/com/microsoft/hyperspace/actions/CreateActionBase.scala

```diff
@@ -166,37 +175,42 @@ private[actions] abstract class CreateActionBase(dataManager: IndexDataManager)
       df: DataFrame,
-      indexConfig: IndexConfig): (DataFrame, Seq[String], Seq[String]) = {
+      indexConfig: IndexConfig,
+      fileIdsMap: Map[String, Long]): (DataFrame, Seq[String], Seq[String]) = {
     val (resolvedIndexedColumns, resolvedIncludedColumns) = resolveConfig(df, indexConfig)
     val columnsFromIndexConfig = resolvedIndexedColumns ++ resolvedIncludedColumns
     val addLineage = indexLineageEnabled(spark)
 
     val indexDF = if (addLineage) {
       // Lineage is captured using two sets of columns:
-      // 1. DATA_FILE_NAME_COLUMN column contains source data filename for each index record.
+      // 1. DATA_FILE_ID_COLUMN column contains source data file id for each index record.
       // 2. If source data is partitioned, all partitioning key(s) are added to index schema
       //    as columns if they are not already part of the schema.
       val missingPartitionColumns = getPartitionColumns(df).filter(
         ResolverUtils.resolve(spark, _, columnsFromIndexConfig).isEmpty)
       val allIndexColumns = columnsFromIndexConfig ++ missingPartitionColumns
 
-      // File path value in DATA_FILE_NAME_COLUMN column is stored as String. We normalize
-      // path values in this column by removing extra preceding `/` characters in the path
-      // and store it the same way paths are stored in Content in an IndexLogEntry instance.
-      // Normalizing path values in DATA_FILE_NAME_COLUMN column keeps path representation
-      // unified in Hyperspace (between index lineage and index metadata) and helps performance
-      // by avoiding the need to fix paths (i.e. removing extra `/` characters) each time
-      // value of DATA_FILE_NAME_COLUMN column is read.
-      // Here is an example of normalization:
+      // File id value in DATA_FILE_ID_COLUMN column is stored as Long.
+      // We lookup the normalized path of the source data file for each record in
+      // fileIdsMap (already generated by Hyperspace when listing source data files
+      // for the index) and store the assigned id to the source data file as the lineage
+      // for the records coming from that source data file.
       // - Original raw path (output of input_file_name() udf, before normalization):
       //    + file:///C:/hyperspace/src/test/part-00003.snappy.parquet
-      // - Normalized path to be stored in DATA_FILE_NAME_COLUMN column:
+      // - Normalized path to be looked up in fileIdsMap:
       //    + file:/C:/hyperspace/src/test/part-00003.snappy.parquet
-      val absolutePath: UserDefinedFunction = udf(
-        (filePath: String) => filePath.replace("file:///", "file:/"))
+      val getFileId: UserDefinedFunction = udf((filePath: String) => {
+        fileIdsMap
+          .get(filePath.replace("file:///", "file:/")) match {
+          case Some(id) => id
+          case _ =>
+            throw HyperspaceException(
+              s"Encountered source data file with unknown file id. (File: $filePath).")
```

Comment:
You mean `Unknown source data found: '$filePath'` since we are looking up by file name?